### PR TITLE
Adding SyncThreads in Cuda to honor dependences

### LINKED ISSUE #48533: JIT-compiled function produces incorrect results
Seems related to broadcasting & TensorExpr fusion.

## To Reproduce

```
import torch

@torch.jit.script
def f(p,a):
	return torch.erfinv(p) * a.sqrt()

p = torch.linspace(0.1,0.9,64,device="cuda")
#p = p.expand(320,-1).contiguous() #not a fix
a = torch.empty(320,1,device="cuda").uniform_(6., 15000.0)
a.requires_grad_() #no error in nograd mode
#p,a = torch.broadcast_tensors(p,a) #fix

r1=f(p,a)
r2=f(p,a)
print(torch.allclose(r1,r2))
print((r1-r2).abs().max())
```

## Environment

 - PyTorch Version (e.g., 1.0): 1.8.0a0 (git)
 - OS (e.g., Linux): windows
 - How you installed PyTorch (`conda`, `pip`, source): source
 - CUDA/cuDNN version: 10.1

cc @gmagogsfm

### PR DESCRIPTION
**Pull request by navahgar**
This PR handles dependences in Cuda by inserting `SyncThreads` in the code as appropriate and necessary.

### CODE REVIEW

File: torch/csrc/jit/tensorexpr/cuda_codegen.cpp [hunk 1]
```diff
@@ -749,7 +751,38 @@ Stmt* GPUMetaVarRewriter::mutate(const For* v) {
...
 }
 
+std::vector<Stmt*> GPUMetaVarRewriter::handleDependences(const Block* v) {
+  analysis::MemDependencyChecker analyzer;
+  v->accept(&analyzer);
+
+  // Iterate over the statements in the block. If any statement is dependent on
+  // any previous statement in the block, then insert a SyncThreads before that
+  // statement. Note that the iteration below is quadratic in the number of
+  // statements in the block.
```
Comment (nickgg) [inline torch/csrc/jit/tensorexpr/cuda_codegen.cpp:761]: I do worry a little about adding an n^2 step here, since it operates on all statements - but I'm not 100% sure how else to do it.

File: torch/csrc/jit/tensorexpr/cuda_codegen.cpp [hunk 2]
```diff
@@ -749,7 +751,38 @@ Stmt* GPUMetaVarRewriter::mutate(const For* v) {
...
+  auto vStmts = v->stmts();
+  std::vector<Stmt*> stmts(vStmts.begin(), vStmts.end());
+  std::set<Stmt*> stmtsNeedingSync;
+  for (auto it1 = stmts.begin(); it1 != stmts.end(); ++it1) {
+    for (auto it2 = std::next(it1); it2 != stmts.end(); ++it2) {
```
Comment (nickgg) [inline torch/csrc/jit/tensorexpr/cuda_codegen.cpp:766]: You probably want it2 to start at it1 to cover the case where there's a hazard between a stmt and the same stmt from another loop iteration.

File: torch/csrc/jit/tensorexpr/cuda_codegen.cpp [hunk 3]
```diff
@@ -749,7 +751,38 @@ Stmt* GPUMetaVarRewriter::mutate(const For* v) {
...
+      auto hazard = getPotentialHazards(analyzer, *it1, *it2);
+      if (hazard != HazardKind::NoDependency) {
+        stmtsNeedingSync.insert(*it2);
+      }
+    }
+  }
+  std::vector<Stmt*> stmtsWithSync;
+  for (size_t i = 0; i < stmts.size(); ++i) {
+    if (stmtsNeedingSync.find(stmts[i]) != stmtsNeedingSync.end()) {
+      stmtsWithSync.push_back(new SyncThreads());
```
Comment (nickgg) [inline torch/csrc/jit/tensorexpr/cuda_codegen.cpp:776]: I think you'll be inserting a lot of false positives here, e.g. 
```
for (int x = 0; x < 10; ++x) {
 A[x] = x;
 B[x] = A[x+1];
}
for (int y = 0; y < 10; ++y) {
 C[y] = A[y+1];
}
```

I think you'll insert two __syncthreads() - once between the store to A[x] and B[x], and once between the two loop stmts. Can you test this case?

File: torch/csrc/jit/tensorexpr/cuda_codegen.cpp [hunk 4]
```diff
@@ -749,7 +751,38 @@ Stmt* GPUMetaVarRewriter::mutate(const For* v) {
...
+  // any previous statement in the block, then insert a SyncThreads before that
+  // statement. Note that the iteration below is quadratic in the number of
+  // statements in the block.
+  auto vStmts = v->stmts();
+  std::vector<Stmt*> stmts(vStmts.begin(), vStmts.end());
+  std::set<Stmt*> stmtsNeedingSync;
+  for (auto it1 = stmts.begin(); it1 != stmts.end(); ++it1) {
+    for (auto it2 = std::next(it1); it2 != stmts.end(); ++it2) {
+      auto hazard = getPotentialHazards(analyzer, *it1, *it2);
+      if (hazard != HazardKind::NoDependency) {
```
Comment (nickgg) [inline torch/csrc/jit/tensorexpr/cuda_codegen.cpp:768]: I think this may lead to false positives: specifically in the case where there are "hazards" that cannot be across threads due to the metavar. E.g.

```
for (int x = 0; x < 10; ++x) { // threadIdx.x
 A[x] = x;
 B[x] = A[x];
}
```

There are no need for syncthreads here. Can you test this case?

**Review by nickgg — CHANGES_REQUESTED** [review]
This looks like it may fix the correctness issues, but I have concerns about performance - both compile time and run time.

It really needs more testing using FileCheck (or similar) to ensure we're not inserting too many `__syncthreads()`.

**Comment by ngimel** [comment]
This and any similar syncthreads-based approach does not (and cannot) save correctness issues, and perf worries are also valid - __syncthreads() are inserted way more often than needed. Let's look at the following example, note the sizes are intentionally big, so that more than one threadblock is needed:
```
@torch.jit.script
def f(p,a):
    tmp = a.sqrt()
    return (tmp,torch.sqrt(p * tmp))

p = torch.linspace(0.1,0.9,64,device="cuda")
a = torch.ones(64,1,device="cuda")
r1=f(p,a)[1]
```
This generates the following kernel:
```
[DEBUG cuda_codegen.cpp:1059] Fused TE CUDA kernel:
[DEBUG cuda_codegen.cpp:1059] 
[DEBUG cuda_codegen.cpp:1059] #define NAN __int_as_float(0x7fffffff)
[DEBUG cuda_codegen.cpp:1059] #define POS_INFINITY __int_as_float(0x7f800000)
[DEBUG cuda_codegen.cpp:1059] #define NEG_INFINITY __int_as_float(0xff800000)
[DEBUG cuda_codegen.cpp:1059] 
[DEBUG cuda_codegen.cpp:1059] 
[DEBUG cuda_codegen.cpp:1059] template<typename T>
[DEBUG cuda_codegen.cpp:1059] __device__ T maximum(T a, T b) {
[DEBUG cuda_codegen.cpp:1059]   return isnan(a) ? a : (a > b ? a : b);
[DEBUG cuda_codegen.cpp:1059] }
[DEBUG cuda_codegen.cpp:1059] 
[DEBUG cuda_codegen.cpp:1059] template<typename T>
[DEBUG cuda_codegen.cpp:1059] __device__ T minimum(T a, T b) {
[DEBUG cuda_codegen.cpp:1059]   return isnan(a) ? a : (a < b ? a : b);
[DEBUG cuda_codegen.cpp:1059] }
[DEBUG cuda_codegen.cpp:1059] 
[DEBUG cuda_codegen.cpp:1059] extern "C" __global__
[DEBUG cuda_codegen.cpp:1059] void fused_sqrt_mul_sqrt(float* t0, float* t1, float* aten_sqrt, float* aten_sqrt_1) {
[DEBUG cuda_codegen.cpp:1059] {
[DEBUG cuda_codegen.cpp:1059]   if (blockIdx.x<1 ? 1 : 0) {
[DEBUG cuda_codegen.cpp:1059]     if (512 * blockIdx.x + threadIdx.x<64 ? 1 : 0) {
[DEBUG cuda_codegen.cpp:1059]       if (blockIdx.x<1 ? 1 : 0) {
[DEBUG cuda_codegen.cpp:1059]         float v = __ldg(t1 + 512 * blockIdx.x + threadIdx.x);
[DEBUG cuda_codegen.cpp:1059]         aten_sqrt_1[512 * blockIdx.x + threadIdx.x] = sqrtf(v);
[DEBUG cuda_codegen.cpp:1059]       }
[DEBUG cuda_codegen.cpp:1059]     }
[DEBUG cuda_codegen.cpp:1059]   }
[DEBUG cuda_codegen.cpp:1059]   __syncthreads();
[DEBUG cuda_codegen.cpp:1059]   float v_1 = __ldg(t0 + (512 * blockIdx.x + threadIdx.x) % 64);
[DEBUG cuda_codegen.cpp:1059]   float v_2 = aten_sqrt_1[(512 * blockIdx.x + threadIdx.x) / 64];
[DEBUG cuda_codegen.cpp:1059]   aten_sqrt[512 * blockIdx.x + threadIdx.x] = sqrtf(v_1 * v_2);
[DEBUG cuda_codegen.cpp:1059] }
[DEBUG cuda_codegen.cpp:1059] }
[DEBUG cuda_codegen.cpp:1059] gpu_block_extents: (8, 1, 1)
[DEBUG cuda_codegen.cpp:1059] gpu_thread_extents: (512, 1, 1)
```
Note how only the 0-th threadblock is writing to `aten_sqrt_1`, but all threadblocks are reading from it! __syncthreads() is solving this problem within a single block (all block 0 writes before `__syncthreads` will be visible for all block 0 reads after `__syncthreads`, but for all other blocks there is absolutely no guarantee that block 0 writes have happened by the time they are reading from aten_sqrt_1. What is needed for this pattern is global sync, but honestly, you don't want to go there, bail out on fusions like these.

**Comment by nickgg** [comment]
@ngimel that is definitely true, but I think we can solve inter-block hazards and intra-block hazards as separate pieces. The example you gave will require us to launch multiple kernels, which we don't yet support.

**Comment by ngimel** [comment]
So what's your plan for detecting and handling inter-block hazards?

**Comment by navahgar** [comment]
Thanks for the comments, @nickgg. Will address them.

Thanks for bringing up the issue with inter-block synchronization. We haven't spent any time on that yet. 

> So what's your plan for detecting and handling inter-block hazards?
If we require multiple kernel launches for all such scenarios (I'm not completely sure about this, but assuming this is the case), wouldn't that (launching separate kernels) take care of the inter-block synchronization as well?

**Comment by ngimel** [comment]
Yes, launching multiple kernels provides device-wide synchronization for writes visibility. An alternative to multiple kernels could be using cooperative groups https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#cooperative-groups.

**Comment by nickgg** [comment]
I had assumed for larger fusion groups we'd need to partition the graph into multiple kernel launches anyway (at least for safe sharing of intermediate memory) and we could just use the same system and take inter-block hazards as necessary cut points. I don't know much about cooperative groups but if they're a better alternative for some cases we should definitely investigate.

I think we'll still need multiple kernel launches though.

File: torch/csrc/jit/tensorexpr/cuda_codegen.cpp [hunk 5]
```diff
@@ -749,7 +751,38 @@ Stmt* GPUMetaVarRewriter::mutate(const For* v) {
   return v->cloneWithNewBody(body);
 }
 
+std::vector<Stmt*> GPUMetaVarRewriter::handleDependences(const Block* v) {
+  analysis::MemDependencyChecker analyzer;
+  v->accept(&analyzer);
```
[the remark this answers is not in the document]
Comment (navahgar) [reply torch/csrc/jit/tensorexpr/cuda_codegen.cpp]: Thanks for pointing that out. Added the dependency handling part as a separate pass, CudaDependencyHandler, independent of the GPUMetaVarRewriter. Now, the MemDependencyAnalyzer will be run on the entire graph and results will be fed into CudaDependencyHandler.

File: torch/csrc/jit/tensorexpr/cuda_codegen.cpp (hunk 1 above)
Comment (navahgar) [reply torch/csrc/jit/tensorexpr/cuda_codegen.cpp]: I don't have an alternative to this as of now. Added a TODO here. Will try to figure something out.

File: torch/csrc/jit/tensorexpr/cuda_codegen.cpp (hunk 2 above)
Comment (navahgar) [reply torch/csrc/jit/tensorexpr/cuda_codegen.cpp]: How are hazards within the same stmt from different iterations meant to be handled? We can't be inserting syncThreads right?

Comment (navahgar) [reply torch/csrc/jit/tensorexpr/cuda_codegen.cpp]: Per our discussion offline, I am not changing this.

File: torch/csrc/jit/tensorexpr/cuda_codegen.cpp (hunk 3 above)
Comment (navahgar) [reply torch/csrc/jit/tensorexpr/cuda_codegen.cpp]: Good point. You were right that my initial implementation was indeed adding two `syncthreads`. 

I have updated the implementation to not add redundant sync threads now, like in this case. PTAL.

File: torch/csrc/jit/tensorexpr/cuda_codegen.cpp (hunk 4 above)
Comment (navahgar) [reply torch/csrc/jit/tensorexpr/cuda_codegen.cpp]: Handle this case as well now.

File: torch/csrc/jit/tensorexpr/cuda_codegen.cpp [hunk 6]
```diff
@@ -661,6 +663,148 @@ class PrioritizeLoad : public IRMutator {
...
+    return new Block(newStmts);
+  }
+
+ private:
+  // Check if the given statements have the same thread index.
+  bool sameThreadIdx(Stmt* s1, Stmt* s2) {
+    auto enclosing_for = [](Stmt* s) -> For* {
+      Stmt* curr = s;
+      while (curr) {
+        if (auto f = dynamic_cast<For*>(curr)) {
```
Comment (nickgg) [inline torch/csrc/jit/tensorexpr/cuda_codegen.cpp:683]: I think this isn't completely correct, check out this case:
```
for (int i = 0; i < 10; i++) {
  for (int j = 0; j < 10; j++) { \\ threadIdx.x
     A[j] += C[i, j];
  }
}
for (int k = 0; k < 10; k++) { \\ threadIdx.x
  B[k] = A[k];
}
```

The two outer loops should pass this check, not fail it. I think you want to find the first enclosing for which has loop options set rather than just the first one.

File: torch/csrc/jit/tensorexpr/loopnest.h [hunk 7]
```diff
@@ -32,13 +32,13 @@ class TORCH_API LoopNest {
   // they can be deduced.
   LoopNest(
       Stmt* stmt,
+      const std::unordered_set<const Buf*>& input_bufs,
```
Comment (nickgg) [inline torch/csrc/jit/tensorexpr/loopnest.h:35]: Can you explain this change? This doesn't seem related. Removing buf_initializers in particular seems somewhat suspicious.

File: torch/csrc/jit/tensorexpr/cuda_codegen.cpp [hunk 8]
```diff
@@ -661,6 +663,148 @@ class PrioritizeLoad : public IRMutator {
...
 };
 
+class CudaDependencyHandler : public IRMutator {
+ public:
+  CudaDependencyHandler(analysis::MemDependencyChecker* checker)
+      : mem_dependency_checker_(checker) {}
+
+  Stmt* mutate(const Block* v) override {
+    auto stmts = v->stmts();
+    auto newStmts = handleDependences({stmts.begin(), stmts.end()});
```
Comment (nickgg) [inline torch/csrc/jit/tensorexpr/cuda_codegen.cpp:673]: const ref instead of copy here?

File: torch/csrc/jit/tensorexpr/cuda_codegen.cpp [hunk 9]
```diff
@@ -661,6 +663,148 @@ class PrioritizeLoad : public IRMutator {
...
+    auto s1Bounds = getInferredBounds(*mem_dependency_checker_, s1, true);
+    auto s2Bounds = getInferredBounds(*mem_dependency_checker_, s2, true);
+    std::unordered_set<const Buf*> syncBuffers;
+    for (const auto& s2Bound : s2Bounds) {
+      const Buf* buf = s2Bound.first;
+      if (s1Bounds.find(buf) == s1Bounds.end()) {
+        continue;
+      }
+
+      // We only handle RAW hazards now.
```
Comment (nickgg) [inline torch/csrc/jit/tensorexpr/cuda_codegen.cpp:716]: What's the reasoning behind this limitation?

File: test/cpp/tensorexpr/test_cuda.cpp [hunk 10]
```diff
@@ -2344,6 +2344,315 @@ TEST(Cuda, MaskMultiDimMultiLevel_CUDA) {
...
+  Stmt* s = l.root_stmt();
+
+  CudaCodeGen cg(s, {a, b, c});
+  std::ostringstream oss;
+  oss << *cg.stmt();
+
+  const std::string& verification_pattern =
+      R"IR(
+        # CHECK: b[threadIdx.x] =
+        # CHECK-NOT: __syncthreads();
```
Comment (nickgg) [inline test/cpp/tensorexpr/test_cuda.cpp:2451]: I'm not super familiar with FileCheck, are you sure this matches ALL lines no containing syncthreads not one line not containing syncthreads?

**Review by nickgg — COMMENTED** [review]
This looks great, nice job @navahgar. A few questions:

Probably also be good if @zheng-xq checked this one as well.

**Comment by navahgar** [comment]
> This looks great, nice job @navahgar. A few questions:
> 
> Probably also be good if @zheng-xq checked this one as well.

Thanks for the review. Requested review from @zheng-xq now.

File: test/cpp/tensorexpr/test_cuda.cpp [hunk 11]
```diff
@@ -2344,6 +2344,315 @@ TEST(Cuda, MaskMultiDimMultiLevel_CUDA) {
...
+  Stmt* s = l.root_stmt();
+
+  CudaCodeGen cg(s, {a, b, c});
+  std::ostringstream oss;
+  oss << *cg.stmt();
+
+  const std::string& verification_pattern =
+      R"IR(
+        # CHECK: b[threadIdx.x] =
+        # CHECK: __syncthreads();
```
Comment (zheng-xq) [inline test/cpp/tensorexpr/test_cuda.cpp:2377]: We need to check that the __syncthreads is outside both loops.

File: test/cpp/tensorexpr/test_cuda.cpp [hunk 12]
```diff
@@ -2344,6 +2344,315 @@ TEST(Cuda, MaskMultiDimMultiLevel_CUDA) {
...
+  s = l.root_stmt();
+
+  CudaCodeGen cg(s, {x, Placeholder(a_buf), Placeholder(b_buf)});
+  std::ostringstream oss;
+  oss << *cg.stmt();
+
+  const std::string& verification_pattern =
+      R"IR(
+        # CHECK: A[threadIdx.x] =
+        # CHECK: __syncthreads();
```
Comment (zheng-xq) [inline test/cpp/tensorexpr/test_cuda.cpp:2414]: I am not entirely sure that this is correct. __syncthreads requires all threads in the block to work. 

It works for this particular example because x is likely to be bound to threadIdx here. So we need to be careful when this is safe and when this is not.

File: test/cpp/tensorexpr/test_cuda.cpp [hunk 13]
```diff
@@ -2344,6 +2344,315 @@ TEST(Cuda, MaskMultiDimMultiLevel_CUDA) {
...
+        # CHECK: b[threadIdx.x] =
+        # CHECK: __syncthreads();
+        # CHECK: c[threadIdx.x] =
+      )IR";
+
+  torch::jit::testing::FileCheck().run(verification_pattern, oss.str());
+}
+
+TEST(Cuda, DependencyInALoop_CUDA) {
+  // for (int x = 0; x < 50; x++) {
```
Comment (zheng-xq) [inline test/cpp/tensorexpr/test_cuda.cpp:2385]: You need to mark "x" as threadIdx here.

**Review by zheng-xq — CHANGES_REQUESTED** [review]
This change is very important. Thank you for taking this up. 

Since handling __syncthreads is a founding block in our system, we want to be cautious here. I will go through more code carefully. But I'd like to see more testing plans for functional correctness. 

1. Run through existing tests and benchmarks, and double check all kernels didn't change. The only exception is that the original kernels were wrong before. 
2. Run through our benchmarks, and make sure there is no performance regressions. 
3. Add more manual tests examples with reductions here.

File: test/cpp/tensorexpr/test_cuda.cpp [hunk 14]
```diff
@@ -2344,6 +2344,315 @@ TEST(Cuda, MaskMultiDimMultiLevel_CUDA) {
...
+  //   A[x] = x;
+  //   B[x] = A[x+1];
```
Comment (zheng-xq) [inline test/cpp/tensorexpr/test_cuda.cpp:2387]: I've actually given this a bit more thoughts. In fact, I would argue that this loop nest is invalid from our IR perspective. 

The major reason is that the default loop semantics is "independent statement evaluation". It is the foundation that we can evaluate each index and its associated statement in parallel, or in arbitrary order. With your example, it has to be evaluated in the serial order. In fact, it will be illegal again if it is "B[x] = A[x - 1]". 

So if in your example: it is first transformed into:

`
for (int x = 0; x < 50; x++) {
    A[x] = x;
}
for (int x = 0; x < 50; x++) {
    B[x] = A[x + 1]
}
`

aka your dependency-across-loop example, that should be fine. 

In summary, I think if the codegen sees the LoopNest in this form, it should be an error. Before that, some transformations should have transformed the LoopNest into a valid form.

**Review by zheng-xq — CHANGES_REQUESTED** [review]

File: torch/csrc/jit/tensorexpr/cuda_codegen.cpp (hunk 6 above)
Comment (navahgar) [reply torch/csrc/jit/tensorexpr/cuda_codegen.cpp:683]: I understand what you are referring to here, in terms of finding the first enclosing for which has loop options set. But I am trying to find an example that needs that case. The example you have results in an error: `"Cannot set both gpu thread and block index"`. Trying to figure out why that is happening.

File: torch/csrc/jit/tensorexpr/cuda_codegen.cpp (hunk 8 above)
Comment (navahgar) [reply torch/csrc/jit/tensorexpr/cuda_codegen.cpp]: Done

File: torch/csrc/jit/tensorexpr/loopnest.h (hunk 7 above)
Comment (navahgar) [reply torch/csrc/jit/tensorexpr/loopnest.h]: I had to add `input_bufs` to get my tests to generate `LoopNest` from arbitrary `For` statement. I was wrong in removing `buf_initializers` as well. Now, I managed the tests without making any changes here. Reverted both the changes.

File: torch/csrc/jit/tensorexpr/cuda_codegen.cpp (hunk 9 above)
Comment (navahgar) [reply torch/csrc/jit/tensorexpr/cuda_codegen.cpp:716]: The reason is the logic we are using to figure out if a buffer has been "synchronized" at a given point. Once we insert a `__syncthreads`, we mark all buffers written so far as "synchronized". 

For WAW or WAR hazards, the buffers will be written more than once. That means that we can not associate `__syncthreads` with a buffer. 

We can handle those by associating `__syncthreads` with `Store`s instead I think. But I wanted to get a simple case working correctly first. I can handle the rest of the cases as follow ups.

File: test/cpp/tensorexpr/test_cuda.cpp (hunk 10 above)
Comment (navahgar) [reply test/cpp/tensorexpr/test_cuda.cpp:2451]: Yeah, FileCheck ensures that `CHECK-NOT` strings do not occur between two matches in this case. More info here: https://llvm.org/docs/CommandGuide/FileCheck.html#the-check-not-directive (I am assuming we are using this FileCheck here.)

File: test/cpp/tensorexpr/test_cuda.cpp (hunk 11 above)
Comment (navahgar) [reply test/cpp/tensorexpr/test_cuda.cpp:2377]: This code is after the loops have been removed, i.e., after `CudaCodeGen`. There are no loops at that point in this example. Am I missing something here?
